src/lazy/streaming_raw_reader.rs

```diff
@@ -325,19 +374,159 @@ impl<SliceType: AsRef<[u8]>> IonDataSource for IonSlice<SliceType> {
             self.buffer()
         );
     }
+
+    fn position(&self) -> usize {
+        self.position
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct IoBuffer {
```

Comment:
🪧 In order for fixed slice inputs and streaming inputs to have a common language for saving input buffers, I reworked things to use this type consistently. `IonSlice` (a fixed input) will avoid constructing an `IoBuffer` as an optimization, but once it's been created it can be re-used by every value in the stream.